stl/inc/xutility

```diff
@@ -4198,7 +4198,368 @@ _NODISCARD _CONSTEXPR17 move_iterator<_Iter> make_move_iterator(_Iter _It) { //
+        }
+    };
+
+public:
+    constexpr common_iterator() noexcept(is_nothrow_constructible_v<_Iter>) // strengthened
+        : _Myiter{}, _Mystate{_State::_Holds_iter} {};
+
+    constexpr common_iterator(_Iter _Right) noexcept(is_nothrow_move_constructible_v<_Iter>) // strengthened
+        : _Myiter{_STD move(_Right)}, _Mystate{_State::_Holds_iter} {}
+
+    constexpr common_iterator(_Se _Right) noexcept(is_nothrow_move_constructible_v<_Se>) // strengthened
+        : _Mysentinel{_STD move(_Right)}, _Mystate{_State::_Holds_sentinel} {}
+
+    constexpr common_iterator(const common_iterator& _Right) noexcept(
+        is_nothrow_copy_constructible_v<const _Iter&>&& is_nothrow_copy_constructible_v<const _Se&>) // strengthened
+        : _Mydummy{}, _Mystate{_Right._Mystate} {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Mystate != _State::_Nothing, "common_iterator can only be assigned if it holds a value");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        if (_Mystate == _State::_Holds_iter) {
+            _Construct_in_place(_Myiter, _Right._Myiter);
+        } else if (_Mystate == _State::_Holds_sentinel) {
+            _Construct_in_place(_Mysentinel, _Right._Mysentinel);
+        }
+    }
+
+    // clang-format off
+    template <class _Iter2, class _Se2>
+        requires convertible_to<const _Iter2&, _Iter> && convertible_to<const _Se2&, _Se>
+    constexpr common_iterator(const common_iterator<_Iter2, _Se2>& _Right) noexcept(
+        is_nothrow_constructible_v<_Iter, const _Iter2&>&& is_nothrow_constructible_v<_Se, const _Se2&>) // strengthened
+        : _Mydummy{}, _Mystate{_Right._Mystate} {
+        // clang-format on
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Mystate != _State::_Nothing, "common_iterator can only be assigned if it holds a value");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        if (_Mystate == _State::_Holds_iter) {
+            _Construct_in_place(_Myiter, _Right._Myiter);
+        } else if (_Mystate == _State::_Holds_sentinel) {
+            _Construct_in_place(_Mysentinel, _Right._Mysentinel);
+        }
+    }
+
+    common_iterator& operator=(const common_iterator& _Right) noexcept(
+        is_nothrow_copy_assignable_v<const _Iter&>&& is_nothrow_copy_assignable_v<const _Se&>&&
+            is_nothrow_copy_constructible_v<const _Iter&>&& is_nothrow_copy_constructible_v<const _Se&>) // strengthened
+    {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Right._Mystate != _State::_Nothing, "common_iterator can only be assigned if it holds a value");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        if (_Mystate == _State::_Holds_iter && _Right._Mystate == _State::_Holds_iter) {
+            if constexpr (!is_nothrow_copy_assignable_v<const _Iter&>) {
+                _Mystate = _State::_Nothing;
+            }
+            _Myiter = _Right._Myiter;
+            if constexpr (!is_nothrow_copy_assignable_v<const _Iter&>) {
+                _Mystate = _State::_Holds_iter;
+            }

```

Comment:
According to [cppreference](https://en.cppreference.com/w/cpp/utility/variant/valueless_by_exception) a variant can become valueless_by_exception in the following situations:

- (guaranteed) an exception is thrown during the move initialization of the contained value from the temporary in copy assignment
- (guaranteed) an exception is thrown during the move initialization of the contained value during move assignment
- (optionally) an exception is thrown when initializing the contained value during a type-changing assignment
- (optionally) an exception is thrown when initializing the contained value during a type-changing emplace

So if copy initialization throws it does indeed become valueless by exception and we need to mark the state accordingly. Am I missing something else?

Reply:
[[variant.assign]/2.3](http://eel.is/c++draft/variant.assign#2.3)

> Otherwise, if `index()` == `j`, assigns the value contained in `rhs` to the value contained in `*this`.

[cppreference](https://en.cppreference.com/w/cpp/utility/variant/operator%3D)

> Otherwise, if `rhs` holds the same alternative as `*this`, assigns the value contained in `rhs` to the value contained in `*this`. If an exception is thrown, `*this` does not become valueless: the value depends on the exception safety guarantee of the alternative's copy assignment.

Reply:
Ah thanks a lot. So i guess we should indeed remove those checks :+1: